Disable virtio-fs for KSM

Jakob-Naucke opened the pull request:
Due to virtio-fs using file-backed memory, KSM (Kernel Samepage Merging) is unstable or, depending on architecture, might not work at all.

- Generalize `.ci/toggle_sandbox_cgroup_only.sh` to `.ci/set_kata_config.sh` so it can be used for other keys.
- Disable virtio-fs for KSM test. Keep virtio-9p if it was enabled.
- Give KSM a little time to kick in (just because there are no shared pages *immediately* after starting container doesn't mean it isn't working).
- Re-enable KSM test on ARM. **I'm assuming this will fix ARM, but I don't have the means to test it. Test before merging.**

Fixes: #3531

fidencio commented:
@Jakob-Naucke, why you don't have how to test it before it gets merged?

Jakob-Naucke commented:
@fidencio sorry, what do you mean? Can't we `/test`/`/test-arm` as usual?

fidencio commented:
> @fidencio sorry, what do you mean? Can't we `/test`/`/test-arm` as usual?

Sorry, I got confused, my bad.  I read your "I'm assuming this will fix ARM, but I don't have the means to test it. Test before merging." as "we don't have means to test it before merging", and I asked myself about "Cant't we `/test-arm` this?" :-)

Anyways, my bad, sorry for the confusion.

cmaf commented:
/test

Jakob-Naucke commented:
On CI failures:

- CLH tests appear to be failing at an earlier stage (Credentials using secrets), so I think it's unrelated?
- ARM failure uncovered a bug: The "do not restore virtio-fs if it was 9p before" functionality is working, but returning 1. Fixed and updated scripts to also print output from the 9p set.
- Metrics failure was because virtio-9p was kept after disabling KSM. Put restore function into `disable_ksm` as well.

Also fixed a subtle logic bug if `restore_virtio_fs` was called without having called `set_ksm_aggressive` and `virtio-9p` was enabled, `virtio-fs` would be set.
Re-requesting review.

GabyCT commented:
/test

GabyCT commented:
Metrics CI is failing with
```
10:08:15 time="2021-05-18T15:08:15Z" level=debug msg="Compare check for [memory-footprint-ksm]"
10:08:15 time="2021-05-18T15:08:15Z" level=debug msg="Checking value [mean]"
10:08:15 time="2021-05-18T15:08:15Z" level=debug msg=" Check minval (102313.860000 > 47667.200000)"
10:08:15 time="2021-05-18T15:08:15Z" level=warning msg="Failed Minval (102313.860000 > 47667.200000) for [memory-footprint-ksm]"
10:08:15 time="2021-05-18T15:08:15Z" level=debug msg=" Check maxval (113083.740000 < 47667.200000)"
10:08:15 time="2021-05-18T15:08:15Z" level=debug msg=Passed
10:08:15 time="2021-05-18T15:08:15Z" level=warning msg="Check for [memory-footprint-ksm] failed [Failed]"
```

Jakob-Naucke commented:
Hmm. Now the metrics failure is a bit surprising to me. I thought it was because it was running on 9p (which it was before) and that was going to be slower, but going by the logs now, some lines later:

```
17:08:15 restoring KSM settings
17:08:15 INFO: already "virtio-fs"
```

Also the static checks were cancelled (?)

GabyCT commented:
@Jakob-Naucke this is the error of static checks
```
GitHub Actions has encountered an internal error when running your job.
```

GabyCT commented:
I open an issue related with github actions https://github.com/actions/virtual-environments/issues/3412

Jakob-Naucke commented:
Digging a bit deeper, yes, the `memory-footprint-ksm` test is supposed to run with KSM (duh), and I think it should run with virtio-9p, even with the performance hit that this implies. @gabyct Do you think the targets could be lowered?

GabyCT commented:
> Digging a bit deeper, yes, the `memory-footprint-ksm` test is supposed to run with KSM (duh), and I think it should run with virtio-9p, even with the performance hit that this implies. @GabyCT Do you think the targets could be lowered?

Yes for me it works fine if we lowered the targets, on the other hand is this PR going to have an impact in https://github.com/kata-containers/tests/blob/main/integration/ksm/ksm_test.sh?

Jakob-Naucke commented:
> is this PR going to have an impact in https://github.com/kata-containers/tests/blob/main/integration/ksm/ksm_test.sh?

@GabyCT well, `ksm_test.sh` will run on ARM again, and virtio-9p will be used for that test...if that answers your question?

fidencio commented:
/test

jodh-intel commented on .ci/set_kata_config.sh:
```diff
@@ -35,18 +46,30 @@ Remove it if you want to use the stateless options.
-current_value=$(kata-runtime kata-env --json | jq ".Runtime.SandboxCgroupOnly")
+if [[ -z "${option}" || ! ${!valid_requests[*]} =~ ${option} ]]
+then
+	# unknown option
```
Why not make this self-documenting?

```bash
echo >&2 "ERROR: unknown option: '$option'"
usage
exit 1
```

Same comment for unknown request below.

Jakob-Naucke replied:
@jodh-intel PTAL again :) maybe this can be simplified further

jodh-intel commented on .ci/set_kata_config.sh:
```diff
@@ -35,18 +46,30 @@ Remove it if you want to use the stateless options.
+fi
+
+# quote non-boolean request
+if [[ "${request}" != "true" && "${request}" != "false" ]]
```
You could simplify this to something like:

```bash
[[ "$request" =~ ^(true|false)$ ]] || request="\"${request}\""
```

Jakob-Naucke replied:
Updated.

jodh-intel commented on metrics/lib/common.bash:
```diff
@@ -227,6 +227,17 @@ set_ksm_aggressive(){
+	# Disable virtio-fs and save whether it was enabled previously
+	set_virtio_out=$(sudo -E "${LIB_DIR}/../../.ci/set_kata_config.sh" shared_fs virtio-9p)
+	echo "${set_virtio_out}"
+	grep "already" <<< "${set_virtio_out}" > /dev/null || was_virtio_fs=true;
```
Slightly shorter to use `grep -q` rather than redirection.

Jakob-Naucke replied:
Updated!

jodh-intel commented on metrics/lib/common.bash:
```diff
@@ -227,6 +227,17 @@ set_ksm_aggressive(){
+
+	# Disable virtio-fs and save whether it was enabled previously
+	set_virtio_out=$(sudo -E "${LIB_DIR}/../../.ci/set_kata_config.sh" shared_fs virtio-9p)
+	echo "${set_virtio_out}"
```
Is this line needed?

Jakob-Naucke replied:
This tells you something like `INFO: already "virtio-fs"` if applicable, I think that should be visible.

jodh-intel commented on metrics/lib/common.bash:
```diff
@@ -273,9 +286,13 @@ wait_ksm_settle(){
 
 	oldscan=$(cat /sys/kernel/mm/ksm/full_scans)
 
+	# Wait some time for KSM to kick in to avoid early dismissal
+	local initial_wait=5
+	sleep ${initial_wait}
+
```
I'm not sure I understand this. We're giving KSM 5 extra seconds to start but then waiting for 5 less seconds for it to settle? Can't we avoid the `sleep` and just increase the timeout in the loop below (or use a timed wait here for KSM to trigger before entering the loop?)

Jakob-Naucke replied:
So without the proposed change, this is run inside the loop:
```sh
newpages=$(cat "${KSM_PAGES_SHARED}")    
[[ "$newpages" -eq 0 ]] && echo "No need to wait for KSM to settle" && return
```

and if KSM hasn't started its work right there and then, this will terminate.
> Can't we [...] use a timed wait here for KSM to trigger before entering the loop?

Isn't that what I'm doing with the `sleep`?

jodh-intel replied:
Unless I'm missing something, ... not really, no ;) You're just sleeping in the "hope" that KSM will start in 5 seconds right? If it takes 6 seconds, the loop below will fail. If it starts in 1 second, the test has to wait a further 4 seconds before it can move to the for loop below.

What I'm saying is can't you use an addional for loop like the one below to poll for some (reasonably large amount of) time for KSM to "start"? Once we know it's started, we can then fall into the "wait for KSM to settle" loop below. By having two loops:

- you can make the logic clearer
- you don't penalise systems where KSM starts quickly as you can jump out of the first loop and hit the second where you wait for KSM to settle.

Jakob-Naucke replied:
Ah, now I get what you meant :) Let me get back.

Jakob-Naucke replied:
Updated.

jodh-intel reviewed (commented):
Thanks @Jakob-Naucke. A few comments.

jodh-intel commented:
Hi @Jakob-Naucke - any update on this?

Jakob-Naucke commented:
Whoops! That was a rebase without the suggested changes.

fidencio commented:
/test

fidencio reviewed (approved):
@Jakob-Naucke, the series look good and clean!

I've noticed the static checks were failing, but the failures didn't seen related at all, thus I just restarted the checks.

Thanks for the work on this.

@jodh-intel, do you want to take a look at the final version? :-)

fidencio commented:
Note to self: We should **only** be used `sandbox_cgroup_only=true`, everywhere. But that's a subject for a different discussion.

fidencio commented:
GitHub action seems to be busted right now.
```
GO111MODULE=on go mod edit -replace=github.com/kata-containers/tests/metrics/exec=../exec
GO111MODULE=on go mod tidy
GO111MODULE=on go mod edit -replace=github.com/kata-containers/tests/metrics/exec=../exec
GO111MODULE=on go mod tidy
go mod tidy: go.mod file indicates go 1.16, but maximum supported version is 1.15
```

Jakob-Naucke commented:
/test

fidencio commented:
/retest-arm
